synapse/storage/databases/main/lock.py

```diff
@@ -103,6 +106,16 @@ async def try_acquire_lock(self, lock_name: str, lock_key: str) -> Optional["Loc
         context manager if the lock is successfully acquired, which *must* be
         used (otherwise the lock will leak).
         """
+        async with self.limiter.queue((lock_name, lock_key)):
+            return await self._try_acquire_lock(lock_name, lock_key)
```

Comment:
This works, though I'm wondering whether a lighter weight approach would be just to have a boolean tracking if we are currently trying to acquire the lock, and if so to immediately return `None`. This has the advantage that we don't needlessly queue up deferreds, but the potential downside is that we'll return `None` *before* we've actually acquired the lock (though I think that is fine).

i.e. something like:

```suggestion
        try:
            if (lock_name, lock_key) in self._acquiring_locks:
                return None
            self._acquiring_locks.add((lock_name, lock_key))
            return await self._try_acquire_lock(lock_name, lock_key)
        finally:
            self._acquiring_locks.discard((lock_name, lock_key))
```

Thoughts?

Reply:
I think that we have to be thread-safe with this, don't we?

Also, I don't think that particular code works since the `finally` block gets run even in the case of early returns which means that after `return None` the key gets removed from `self._acquiring_locks`.

Reply:
```suggestion
        if (lock_name, lock_key) in self._acquiring_locks:
            return None
        try:
            self._acquiring_locks.add((lock_name, lock_key))
            return await self._try_acquire_lock(lock_name, lock_key)
        finally:
            self._acquiring_locks.discard((lock_name, lock_key))
```

I think this is right? It will guarantee that it will always discard the key after `self._try_acquire_lock`, but in the case where it returns `None`, we don't discard the lock.

Reply:
Thinking about this more, I’m not sure what we want the behavior to be. We could make it so that it blocks until it can acquire a lock (that is what I have implemented currently) or we can make it fail to return a lock when someone else has the lock. I’m not really sure what the best option is.

Reply:
Lock APIs usually have the option of blocking until the lock is acquired, or giving up immediately if someone else holds the lock. eg. https://docs.python.org/3/library/threading.html#threading.Lock.acquire

I think it should be the latter, since the method is named `try_acquire_lock` rather than `acquire_lock` (which we haven't implemented).